Bounds and proj

bbuzz31 opened the pull request:
This updates the bounding box based on discussion with @piyushrpt.

Conversions are ugly but quick first cut to check that setup is correct @dbekaert @jlmaurer 

It's just the last commit (calculate ray tracing buffer...)

dbekaert commented on tools/RAiDER/utilFcns.py:
```diff
@@ -1222,3 +1223,36 @@ def get_dt(t1,t2):
+        lon, lat = transform_coords(4087, 4326, lon, lat)
+
+    # get the top of the model
+    dct_tom = {'HRES': 80301.65, 'HRRR': 80301.65, 'GMAO': 80301.65, 'ERA5': 80301.65}
+    tom     = dct_tom[model]
```
this we should likely put in the model class itself such as HRRR or HRES?

dbekaert commented on tools/RAiDER/utilFcns.py:
```diff
@@ -1222,3 +1223,36 @@ def get_dt(t1,t2):
+    # get the top of the model
+    dct_tom = {'HRES': 80301.65, 'HRRR': 80301.65, 'GMAO': 80301.65, 'ERA5': 80301.65}
+    tom     = dct_tom[model]
+    near_range = tom / np.sin(np.deg2rad(30))
```
This should be drive by the angle at near range. This will be much overkill for UAVSAr for padding, so perhaps the 30 degrees is good enough for now.

dbekaert commented on tools/RAiDER/utilFcns.py:
```diff
@@ -1222,3 +1223,36 @@ def get_dt(t1,t2):
+    tom     = dct_tom[model]
+    near_range = tom / np.sin(np.deg2rad(30))
+
+    # this is in meters tho
```
clean up language "this is in meters"

dbekaert commented on tools/RAiDER/utilFcns.py:
```diff
@@ -1222,3 +1223,36 @@ def get_dt(t1,t2):
+        proj = False
+    else:
+        proj = True
+        lon, lat = transform_coords(4087, 4326, lon, lat)
```
This is a hard coded assumption. dont do that.
Are these lon lat from the user? are they always in lonlat? if they are not cant we carry the proj with it and do it correct.?

dbekaert commented on tools/RAiDER/utilFcns.py:
```diff
@@ -1222,3 +1223,36 @@ def get_dt(t1,t2):
+        lon, lat = transform_coords(4087, 4326, lon, lat)
+
+    # get the top of the model
+    dct_tom = {'HRES': 80301.65, 'HRRR': 80301.65, 'GMAO': 80301.65, 'ERA5': 80301.65}
```
This one should really exist for each model that we support. 
ERA5T, ERA-I, NCUM, MERRA etc otehrwize this will error in future.

dbekaert commented on tools/RAiDER/utilFcns.py:
```diff
@@ -1222,3 +1223,36 @@ def get_dt(t1,t2):
+        proj = True
+        lon, lat = transform_coords(4087, 4326, lon, lat)
+
+    # get the top of the model
```
Also capture that this is in meters in spherical projection?

dbekaert commented on tools/RAiDER/utilFcns.py:
```diff
@@ -1222,3 +1223,36 @@ def get_dt(t1,t2):
+    # this is in meters tho
+    buffer   = near_range / np.cos(np.deg2rad(lat))
+
+    x, y = transform_coords(4326, 4087, lon, lat)
+    poly = Point(x, y).buffer(buffer)
```
based on conversation what the lon lat is the centroid of the ROI. You are now using the centroid and adding buffer. So you missed now half of the ROI.

dbekaert replied:
The buffer you need to add to the extends of the ROI

dbekaert replied:
So i think you need to have the BOX and then transform it to a local coordinate system in m; then path it with the buffer; and covert it back to the source proj; that would be the new ROI.

dbekaert commented on CHANGELOG.md:
```diff
@@ -7,6 +7,8 @@ and this project adheres to [PEP 440](https://www.python.org/dev/peps/pep-0440/)
 and uses [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [0.4.3]
++ Error out of RAiDER when fetching weather model if request to API fails; previously would skip datetime and continue
```
I think this should be reverted. Only error out in the workflow that is called by HYP3.

dbekaert commented:
Can you also add in an examples when you are doing to confirm its working. e.g. for equator and the high latitude.

bbuzz31 commented:
![image](https://user-images.githubusercontent.com/15522503/226477781-4d4630f6-2fe8-44b1-94fc-c7486c351638.png)

difference in total delay  between projected and ray (units = meters) after my fix (not yet pushed)
from left: projected delays (m); ray tracing delays (m); residual
all HRES model levels

bbuzz31 commented:
Here are a couple test cases:
```
  Original S/N/W/E: [79, 80.5, -117, 116]
  Updated  S/N/W/E: [70.26, 89.24, -125.74, 124.74]
  
  Original S/N/W/E: [0, 2, -117, 116]
  Updated  S/N/W/E: [-1.44, 3.44, -118.44, 117.44]
  
  Original S/N/W/E: [34, 36, -117, 116]
  Updated  S/N/W/E: [32.22, 37.78, -118.78, 117.78]

dbekaert commented on test/test_temporal_interpolate.py:
```diff
@@ -68,7 +69,7 @@ def test_cube_timemean():
     date       = 20200130
     dct_hrs    = {'GMAO': [12, 15, '13:30:00'], 'ERA5': [13, 14, '13:30:00']}
-    hr1, hr2, ti = dct_hrs[WM]
+    hr1, hr2, ti = dct_hrs[wm]
```
i noticed the changing from WM and wm back and forth. Rationale?

bbuzz31 replied:
Some files are called 'ERA-5' and others are 'ERA5'. I probably have some wrong somewhere still; will update

dbekaert replied:
Yes, this seems like a code hack to make it work rather than going to the root of the cause.

bbuzz31 replied:
it's just the weather model file name

bbuzz31 commented on test/test_intersect.py:
```diff
@@ -87,18 +89,19 @@ def test_cube_intersect():
-    wm_file = os.path.join(SCENARIO_DIR, f'{WM}_hydro_{date}T{time.replace(":", "")}_ztd.tiff')
-    da      = xr.open_dataset(wm_file)['band_data']
-    gold    = {'GMAO': 2.045914, 'ERA5': 2.061974, 'HRRR': 3.0972726}
+    WM_file = os.path.join(SCENARIO_DIR, f'{WM}_hydro_{date}T{time.replace(":", "")}_ztd.tiff')
+    da      = xr.open_dataset(WM_file)['band_data']
+    gold    = {'GMAO': 2.0541468, 'ERA5': 2.0696816}#, 'HRRR': 3.0972726}
```
we now produce data on a cube of different shape; these are the mean values so are updated accordingly

dbekaert replied:
Can you elaborate more. Yes the cube and buffering has changed, are you not computing these values for a fixed lon lat? In that case it would not be impacted by the buffering.

dbekaert replied:
We should take a single value, can be seperate PR.

dbekaert commented:
> Here are a couple test cases:
> 
> ```
>   Original S/N/W/E: [79, 80.5, -117, 116]
>   Updated  S/N/W/E: [70.26, 89.24, -125.74, 124.74]
>   
>   Original S/N/W/E: [0, 2, -117, 116]
>   Updated  S/N/W/E: [-1.44, 3.44, -118.44, 117.44]
>   
>   Original S/N/W/E: [34, 36, -117, 116]
>   Updated  S/N/W/E: [32.22, 37.78, -118.78, 117.78]
> ```

Show cases where you do -180 or 90 degrees cross overs too.

jlmaurer commented on tools/RAiDER/cli/raider.py:
```diff
@@ -214,9 +214,13 @@ def calcDelays(iargs=None):
+
         else:
-            ll_bounds = aoi.add_buffer(buffer=1)
+            ll_bounds = aoi.add_buffer(buffer=0.05)
```
I think this is outdated.

dbekaert commented on CHANGELOG.md:
```diff
@@ -7,6 +7,12 @@ and this project adheres to [PEP 440](https://www.python.org/dev/peps/pep-0440/)
+    + then, if ray tracing is used, buffer based on the 
+    + then if ray tracing us used, to pad the downloaded model that is used for computing the 3D cube. Assumes look angle is fixed increases with latitude.
+       
++ Update and convert user given AOI to weather model projection (except for HRRR)
```
This is not correct. Its not model specific. its either SNWE or its user coordinate file with projection. dont think following our dicussion this is still needed.

jlmaurer replied:
I should have updated this; I think this line is not needed anymore?

dbekaert commented on CHANGELOG.md:
```diff
@@ -7,6 +7,12 @@ and this project adheres to [PEP 440](https://www.python.org/dev/peps/pep-0440/)
 
 ## [0.4.3]
++ Add two stage buffer; 
+    + first pad user bounding box such that a 3D cube isgenerated that at min covers user area of interest.
```
typo

dbekaert commented on tools/RAiDER/llreader.py:
```diff
@@ -53,27 +55,69 @@ def projection(self):
+
+        # buffer on the side nearest the sensor
+        if ((lookDir == 'right') and (direction == 'asc')) or ((lookDir == 'left') and (direction == 'desc')):
+            W = W - buffer
+        else:
+            E = E + buffer
```
Same comment here about date time crossing as you had for the other buffer

jlmaurer replied:
@dbekaert @bbuzz31 On this topic it may be good to do some creative thinking, e.g., leaving longitudes unwrapped for -180-180 and converting to 0-360 when needed in order to avoid having two sub-AOIs defined (one >-180 and one <180). Didn't go too far with that thinking here, thought it may need its own PR.

bbuzz31 replied:
I agree, this should be a separate PR

dbekaert commented on tools/RAiDER/losreader.py:
```diff
@@ -197,6 +198,23 @@ def __init__(self, filename=None, los_convention='isce', time=None, look_dir = '
             raise RuntimeError(f"Unknown look direction: {look_dir}")
+    
+
+    def getSensorDirection(self):
+        if self._orbit is None:
+            raise ValueError('The orbit has not been set')
+        z = self._orbit.position[:,2]
+        t = self._orbit.time
+        start = np.argmin(t)
+        end = np.argmax(t)
+        if z[start] > z[end]:
+            return 'desc'
+        else:
+            return 'asc'
```
Does this not make ray-tracing break in case the LOS is provided as curbed. We plan to add that in future.

jlmaurer replied:
> Does this not make ray-tracing break in case the LOS is provided as curbed. We plan to add that in future.

@dbekaert what is "curbed"? I'm not familiar with that term. This code relies on the fact that the orbit object provides x/y/z position for the sensor. As long as you have the orbit object you should have the x/y/z coordinates.

bbuzz31 replied:
That was a typo. He meant providing an LOS cube. Yes, this will break raytracing @dbekaert.
I have added this to the issue ticket regarding the LOS cube and we will have to address it when it's implemented, but for now I think this PR should be merged.

jlmaurer commented on tools/RAiDER/delay.py:
```diff
@@ -105,7 +105,7 @@ def tropo_delay(
             ifWet, ifHydro = getInterpolators(ds, "ztd")
         except RuntimeError:
-            logger.exception('Weather model {} failed, may contain NaNs'.format(weather_model_file))
+            logger.exception('Weather model %s failed, may contain NaNs', weater_model_file)
```
I think this is still a typo.

dbekaert commented:
What if los files are provided as cubes and not the orbit?

On Sun, Mar 26, 2023 at 5:10 PM Jeremy Maurer ***@***.***>
wrote:

> ***@***.**** approved this pull request.
> ------------------------------
>
> In tools/RAiDER/delay.py
> <https://github.com/dbekaert/RAiDER/pull/506#discussion_r1144146056>:
>
> > @@ -105,7 +105,7 @@ def tropo_delay(
>          try:
>              ifWet, ifHydro = getInterpolators(ds, "ztd")
>          except RuntimeError:
> -            logger.exception('Weather model {} failed, may contain NaNs'.format(weather_model_file))
> +            logger.exception('Weather model %s failed, may contain NaNs', weater_model_file)
>
> Typo?
> ------------------------------
>
> In tools/RAiDER/llreader.py
> <https://github.com/dbekaert/RAiDER/pull/506#discussion_r1144187849>:
>
> > +                project  = Transformer.from_crs(self._proj, wm_proj, always_xy=True).transform
> +                aoi_proj = transform(project, aoi_orig)
> +            else:
> +                raise Exception('Unsupported transformation of {self.type()} AOI.')
> +
> +            self._proj = wm_proj
> +        return
> +
> +
> +    def add_buffer(self, model:object):
> +        """ Add a fixed buffer to the AOI. Assume weather model is in wgs84.
> +
> +        Ensures cube is slighly larger than requested area.
> +        """
> +        wm_proj  = CRS.from_epsg(4326)
> +        buffer   = 0.5 # degrees; should eventually be model specific
>
> One thing we should keep in mind here is that the buffered extents are not
> currently being rounded, so if e.g. the extent of the GUNW changes slightly
> this could result in the garnered nodes changing.
> ------------------------------
>
> In CHANGELOG.md
> <https://github.com/dbekaert/RAiDER/pull/506#discussion_r1148679130>:
>
> > @@ -7,6 +7,12 @@ and this project adheres to [PEP 440](https://www.python.org/dev/peps/pep-0440/)
>  and uses [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
>
>  ## [0.4.3]
> ++ Add two stage buffer;
> +    + first pad user bounding box such that a 3D cube isgenerated that at min covers user area of interest.
> +    + then, if ray tracing is used, buffer based on the
> +    + then if ray tracing us used, to pad the downloaded model that is used for computing the 3D cube. Assumes look angle is fixed increases with latitude.
> +
> ++ Update and convert user given AOI to weather model projection (except for HRRR)
>
> I should have updated this; I think this line is not needed anymore?
> ------------------------------
>
> In tools/RAiDER/delay.py
> <https://github.com/dbekaert/RAiDER/pull/506#discussion_r1148679315>:
>
> > @@ -105,7 +105,7 @@ def tropo_delay(
>          try:
>              ifWet, ifHydro = getInterpolators(ds, "ztd")
>          except RuntimeError:
> -            logger.exception('Weather model {} failed, may contain NaNs'.format(weather_model_file))
> +            logger.exception('Weather model %s failed, may contain NaNs', weater_model_file)
>
> I think this is still a typo.
> ------------------------------
>
> In tools/RAiDER/llreader.py
> <https://github.com/dbekaert/RAiDER/pull/506#discussion_r1148679940>:
>
> > +            W = W - buffer
> +        else:
> +            E = E + buffer
>
> @dbekaert <https://github.com/dbekaert> @bbuzz31
> <https://github.com/bbuzz31> On this topic it may be good to do some
> creative thinking, e.g., leaving longitudes unwrapped for -180-180 and
> converting to 0-360 when needed in order to avoid having two sub-AOIs
> defined (one >-180 and one <180). Didn't go too far with that thinking
> here, thought it may need its own PR.
> ------------------------------
>
> In tools/RAiDER/losreader.py
> <https://github.com/dbekaert/RAiDER/pull/506#discussion_r1148678899>:
>
> > +    def getSensorDirection(self):
> +        if self._orbit is None:
> +            raise ValueError('The orbit has not been set')
> +        z = self._orbit.position[:,2]
> +        t = self._orbit.time
> +        start = np.argmin(t)
> +        end = np.argmax(t)
> +        if z[start] > z[end]:
> +            return 'desc'
> +        else:
> +            return 'asc'
>
> Does this not make ray-tracing break in case the LOS is provided as
> curbed. We plan to add that in future.
>
> @dbekaert <https://github.com/dbekaert> what is "curbed"? I'm not
> familiar with that term. This code relies on the fact that the orbit object
> provides x/y/z position for the sensor. As long as you have the orbit
> object you should have the x/y/z coordinates.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dbekaert/RAiDER/pull/506#pullrequestreview-1351603580>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AESZPSMLV7YTUBAQGJO7AYLW6DLHTANCNFSM6AAAAAAV7EWM3Q>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

jlmaurer commented on tools/RAiDER/llreader.py:
```diff
@@ -53,27 +55,97 @@ def projection(self):
+        Ensures cube is slighly larger than requested area.
+        """
+        wm_proj  = CRS.from_epsg(4326)
+        buffer   = 0.5 # degrees; should eventually be model specific
```
@bbuz yes, I introduced rounding in order to keep bounding box numbers nice.

bbuzz31 commented:
> > Here are a couple test cases:
> > ```
> >   Original S/N/W/E: [79, 80.5, -117, 116]
> >   Updated  S/N/W/E: [70.26, 89.24, -125.74, 124.74]
> >   
> >   Original S/N/W/E: [0, 2, -117, 116]
> >   Updated  S/N/W/E: [-1.44, 3.44, -118.44, 117.44]
> >   
> >   Original S/N/W/E: [34, 36, -117, 116]
> >   Updated  S/N/W/E: [32.22, 37.78, -118.78, 117.78]
> > ```
> 
> Show cases where you do -180 or 90 degrees cross overs too.

I think this should be reserved for another PR

bbuzz31 commented:
![Screenshot 2023-03-29 at 15 23 00](https://user-images.githubusercontent.com/15522503/228698934-41d0fb54-d152-4390-81f7-e323c88460d7.png)

Here are the projected/ray/projected-ray  maps in meters

bbuzz31 commented:
> > > Here are a couple test cases:
> > > ```
> > >   Original S/N/W/E: [79, 80.5, -117, 116]
> > >   Updated  S/N/W/E: [70.26, 89.24, -125.74, 124.74]
> > >   
> > >   Original S/N/W/E: [0, 2, -117, 116]
> > >   Updated  S/N/W/E: [-1.44, 3.44, -118.44, 117.44]
> > >   
> > >   Original S/N/W/E: [34, 36, -117, 116]
> > >   Updated  S/N/W/E: [32.22, 37.78, -118.78, 117.78]
> > > ```
> > 
> > 
> > Show cases where you do -180 or 90 degrees cross overs too.
> 
> I think this should be reserved for another PR

With @jlmaurer simplification and updated equation (thanks!), we now get the following for ascending, right looking
```
Original S/N/W/E: [79, 80.5, -10, 10]
Ray trace buffer  S/N/W/E: [79, 80.5, -12.8, 10]
Buffered west by: 2.80º

Original S/N/W/E: [34, 36, -10, 10]
Ray trace buffer  S/N/W/E: [34, 36, -10.57, 10]
Buffered west by: 0.57º

Original S/N/W/E: [0, 2, -10, 10]
Ray trace buffer  S/N/W/E: [0, 2, -10.46, 10]
Buffered west by: 0.46º

Original S/N/W/E: [-34, -36, -10, 10]
Ray trace buffer  S/N/W/E: [-34, -36, -10.57, 10]
Buffered west by: 0.57º

Original S/N/W/E: [-85, -89, -10, 10]
Ray trace buffer  S/N/W/E: [-85, -89, -36.47, 10]
Buffered west by: 26.47º
```

If the resulting buffer extends past +/- 180, we now raise a warning. Will address this in a future PR.

jlmaurer commented:
@bbuzz31 is this ready to merge? I see there is a unit test failing. Also do you know why the ray-std plot is banded?

bbuzz31 commented:
> @bbuzz31 is this ready to merge? I see there is a unit test failing. Also do you know why the ray-std plot is banded?

No waiting on response from @dbekaert.
Not sure why test failed, just had to rerun for it to pass.
I think we're seeing the lateral variation that's picked up by the ray tracing but not the zenith-> slant

jlmaurer commented:
> I think we're seeing the lateral variation that's picked up by the ray tracing but not the zenith-> slant

Super interesting. So then the banding would be related to the horizontal size required for the rays to reach the top of the atmosphere?
